nixos/modules/tasks/cpu-freq.nix

```diff
@@ -28,25 +49,37 @@ in
 
   ###### implementation
 
-  config = mkIf (!config.boot.isContainer && config.powerManagement.cpuFreqGovernor != null) {
+  config =
+    let
+      governorSet = cfg.governor != null;
+      maxSet = cfg.max != null;
+      minSet = cfg.min != null;
+      shouldEnable =
+        !config.boot.isContainer &&
+        (governorSet || maxSet || minSet);
+    in
+    mkIf shouldEnable {
 
-    boot.kernelModules = [ "cpufreq_${cfg.cpuFreqGovernor}" ];
+      boot.kernelModules = optional governorSet "cpufreq_${cfg.governor}";
 
-    environment.systemPackages = [ cpupower ];
+      environment.systemPackages = [ cpupower ];
 
-    systemd.services.cpufreq = {
-      description = "CPU Frequency Governor Setup";
-      after = [ "systemd-modules-load.service" ];
-      wantedBy = [ "multi-user.target" ];
-      path = [ cpupower pkgs.kmod ];
-      unitConfig.ConditionVirtualization = false;
-      serviceConfig = {
-        Type = "oneshot";
-        RemainAfterExit = "yes";
-        ExecStart = "${cpupower}/bin/cpupower frequency-set -g ${cfg.cpuFreqGovernor}";
-        SuccessExitStatus = "0 237";
+      systemd.services.cpufreq = {
+        description = "CPU Frequency Setup";
+        after = [ "systemd-modules-load.service" ];
+        wantedBy = [ "multi-user.target" ];
+        path = [ cpupower pkgs.kmod ];
+        unitConfig.ConditionVirtualization = false;
+        serviceConfig = {
+          Type = "oneshot";
+          RemainAfterExit = "yes";
+          ExecStart =
+            optional governorSet "${cpupower}/bin/cpupower frequency-set --governor ${cfg.governor}" ++
+            optional maxSet "${cpupower}/bin/cpupower frequency-set --max ${toString cfg.max}" ++
+            optional minSet "${cpupower}/bin/cpupower frequency-set --min ${toString cfg.min}";
```

Comment:
Do something like this instead:
```nix
let
  enable = cfg.governor != null;
  # ...
in {
  # ...
  config = mkIf enable {
    # ...
    ExecStart = "${cpupower}/bin/cpupower frequency-set --governor ${governor}"
    + optionalString (cfg.min != null) " frequency-set --min ${toString cfg.min}"
    + optionalString (cfg.max != null) " frequency-set --max ${toString cfg.max}";
  };
}
```

Reply:
@Infinisil Thanks for your fast review!

It looks like you are suggesting making the following three changes:

1. Use `optionalString` instead of having multiple `ExecStart`.
1. In the `enable = cfg.governor != null` line, it looks like you are suggesting to only enable the systemd service if the governor is set, and not depend on whether the whether or not `cfg.max` or `cfg.min` are set.
1. In the `enable = cfg.governor != null` line, it looks like you are suggesting to only enable the systemd service if the governor is set, and not check whether or not we are in a container.

I'm pretty sure you meant 1, but I'm not sure if you actually intended for me to change 2 and 3 as well.  In any case, let me respond to all three below:

1. Using `optionalString` is definitely a good idea!  I fixed this in commit 91fe342.
1. It is actually possible to set the max and min CPU frequency even if the governor is not set.  That is the reason the systemd service is enabled if ANY OF the max, min, or governor is set.  I think this should be kept as-is.
1. Currently, the cpufreq systemd service is only created if you are not running in a container.  This PR keeps this functionality.  I'm not sure what the right thing to do here is, so I just went with how it currently works.

Reply:
Oh I see, thanks for the clarification, sorry I didn't get it initially.